libafl_qemu/src/hooks.rs

```diff
@@ -0,0 +1,860 @@
+//! The high-level hooks
+use core::{
+    ffi::c_void,
+    fmt::{self, Debug, Formatter},
+    marker::{PhantomData, PhantomPinned},
+    mem::transmute,
+    pin::Pin,
+    ptr::{self, addr_of},
+};
+
+use libafl::{executors::inprocess::inprocess_get_state, inputs::Input};
+
+pub use crate::emu::SyscallHookResult;
+use crate::{
+    emu::{Emulator, SKIP_EXEC_HOOK},
+    helper::{QemuHelper, QemuHelperTuple},
+    GuestAddr,
+};
+
+static mut QEMU_HELPERS_PTR: *const c_void = ptr::null();
+unsafe fn get_qemu_helpers<'a, QT>() -> &'a mut QT {
+    (QEMU_HELPERS_PTR as *mut QT)
+        .as_mut()
+        .expect("A high-level hook is installed but QemuHooks is not initialized")
+}
+
+static mut GEN_EDGE_HOOK_PTR: *const c_void = ptr::null();
+extern "C" fn gen_edge_hook_wrapper<I, QT, S>(src: u64, dst: u64) -> u64
+where
+    I: Input,
+    QT: QemuHelperTuple<I, S>,
+{
+    unsafe {
+        let helpers = get_qemu_helpers::<QT>();
+        let emulator = Emulator::new_empty();
+        let func: fn(&Emulator, &mut QT, Option<&mut S>, u64, u64) -> Option<u64> =
+            transmute(GEN_EDGE_HOOK_PTR);
+        (func)(&emulator, helpers, inprocess_get_state::<S>(), src, dst)
+            .map_or(SKIP_EXEC_HOOK, |id| id)
+    }
+}
+
+static mut EDGE_HOOKS: Vec<*const c_void> = vec![];
+extern "C" fn edge_hooks_wrapper<I, QT, S>(id: u64)
+where
+    I: Input,
+    QT: QemuHelperTuple<I, S>,
+{
+    let helpers = unsafe { get_qemu_helpers::<QT>() };
+    let emulator = Emulator::new_empty();
+    for hook in unsafe { &EDGE_HOOKS } {
+        let func: fn(&Emulator, &mut QT, Option<&mut S>, u64) = unsafe { transmute(*hook) };
+        (func)(&emulator, helpers, inprocess_get_state::<S>(), id);
+    }
+}
+
+static mut GEN_BLOCK_HOOK_PTR: *const c_void = ptr::null();
+extern "C" fn gen_block_hook_wrapper<I, QT, S>(pc: u64) -> u64
+where
+    I: Input,
+    QT: QemuHelperTuple<I, S>,
+{
+    unsafe {
+        let helpers = get_qemu_helpers::<QT>();
+        let emulator = Emulator::new_empty();
+        let func: fn(&Emulator, &mut QT, Option<&mut S>, u64) -> Option<u64> =
+            transmute(GEN_EDGE_HOOK_PTR);
```

Comment:
I discovered this bug just now. Can you include this small fix while you are already moving the hooks?
```suggestion
            transmute(GEN_BLOCK_HOOK_PTR);
```

Reply:
sure ty